### ISSUE TITLE
consistent type for genericBuild phases

### ISSUE THREAD

**Issue by teto** [NixOS/nixpkgs#31682]
## Issue description
A issue that hit me quite a few times is that most docs refer to the build phases in a nix-shell via
patchPhase/configurePhase etc. Sometimes I got surprised that configurePhase would do nothing. Only recently did I realize (as a sideeffect of someone telling me about `set|less`) that sometimes  `$configurePhase` was fine while sometimes it should be `$eval "$configurePhase"`.
Isn't there a way to always be able to use the short form `$configurePhase` ? like in bash unset  the default configurePhase and redefine it with the configurePhase string instead ? or some other workaround. This would reduce the cognitive load and prevent rookie misunderstandings.

**Comment by teto** [comment]
Also it would allow to plug in third party tools such as neomake so that it would just have to run "buildPhase" (relevant discussion https://github.com/neomake/neomake/issues/1594)

I commented on the mailing list too: https://groups.google.com/forum/#!topic/nix-devel/FdrsuYeQk6Q

**Comment by danbst** [comment]
Yeah, we should collect all issues when build behavior isn't matched with building inside nix-shell

I know of two other things:
1. unpackPhase in builder makes `cd ..` into unpacked folder, while in nix-shell it doesn't
2. nix-shell doesn't set all build hooks. So `configurePhase` or even `eval "$configurePhase"` doesn't work for some build systems (CMake, I think)

**Comment by teto** [comment]
On a related cool rfc https://github.com/NixOS/rfcs/pull/32

**Comment by teto** [comment]
very much interested still

**Comment by arafangion** [comment]
I was going *nuts* trying to figure this out, until someone pointed out this bug for me, and pointed out https://github.com/NixOS/nix/pull/758 which I've now also left a comment on.
